File: TextosJapones/src/main/java/org/jisho/textosJapones/controller/mangas/MangasComicInfoMalId.java
package org.jisho.textosJapones.controller.mangas;

import com.jfoenix.controls.JFXButton;
import com.jfoenix.controls.JFXDialog;
import com.jfoenix.controls.JFXDialogLayout;
import com.jfoenix.controls.JFXTextField;
import com.jfoenix.controls.events.JFXDialogEvent;
import javafx.event.EventHandler;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.fxml.Initializable;
import javafx.geometry.Insets;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.control.Label;
import javafx.scene.effect.BoxBlur;
import javafx.scene.image.ImageView;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.StackPane;
import javafx.scene.paint.Color;
import javafx.scene.robot.Robot;
import javafx.scene.text.Font;
import javafx.util.Callback;
import org.jisho.textosJapones.components.notification.AlertasPopup;
import org.jisho.textosJapones.model.entities.comicinfo.MAL;
import org.jisho.textosJapones.model.entities.comicinfo.MAL.Registro;
import org.jisho.textosJapones.processar.comicinfo.ProcessaComicInfo;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.ResourceBundle;

public class MangasComicInfoMalId implements Initializable {

	private static final Logger LOGGER = LoggerFactory.getLogger(MangasComicInfoMalId.class);

	final private static String STYLE_SHEET = AlertasPopup.class.getResource("/css/Dark_Theme.css").toExternalForm();

	@FXML
	public JFXTextField txtId;

	@FXML
	public JFXTextField txtNome;

	@FXML
	public ImageView imagem;
	
	private final MAL mal = new MAL("", "");
	private final Robot robot = new Robot();
	private Registro consulta = null;

	private static JFXButton btnConfirmar;
	private static JFXButton btnVoltar;
	private static JFXDialog dialog;
	
	private Boolean validaCampos() {
		Boolean valida = !txtId.getText().replaceAll("/[^0-9]+/g", "").trim().isEmpty();
		if (!valida)
			txtId.setUnFocusColor(Color.RED);
		return valida;
	}
	
	public String getId() {
		return validaCampos() ? txtId.getText() : "";
	}
	
	public Registro getObjeto() {
		return consulta;
	}
	
	public static void abreTelaCorrecao(StackPane rootStackPane, Node nodeBlur, Callback<Registro, Boolean> callback) {
		try {
			BoxBlur blur = new BoxBlur(3, 3, 3);
			JFXDialogLayout dialogLayout = new JFXDialogLayout();
			dialog = new JFXDialog(rootStackPane, dialogLayout, JFXDialog.DialogTransition.CENTER);

			FXMLLoader loader = new FXMLLoader();
			loader.setLocation(getFxmlLocate());
			Parent newAnchorPane = loader.load();
			MangasComicInfoMalId cnt = loader.getController();

			Label titulo = new Label("Troca de Id");
			titulo.setFont(Font.font(20));
			titulo.setTextFill(Color.web("#ffffff", 0.8));

			List<JFXButton> botoes = new ArrayList<JFXButton>();

			btnConfirmar = new JFXButton("Confirmar");
			btnConfirmar.setOnAction(AE -> {
				String numero = cnt.getId().replaceAll("/[^0-9]+/g", "");
				if (!numero.isEmpty()) {
					callback.call(cnt.getObjeto());
					dialog.close();
				}
			});
			btnConfirmar.getStyleClass().add("background-Green2");
			botoes.add(btnConfirmar);
			
			btnVoltar = new JFXButton("Voltar");
			btnVoltar.setOnAction(AV -> dialog.close());
			btnVoltar.getStyleClass().add("background-White1");
			botoes.add(btnVoltar);

			dialogLayout.setHeading(titulo);
			dialogLayout.setBody(newAnchorPane);
			dialogLayout.setActions(botoes);

			dialog.getStylesheets().add(STYLE_SHEET);
			dialog.setPadding(new Insets(0, 0, 0, 0));
			dialog.setOnDialogClosed((JFXDialogEvent event1) -> {
				nodeBlur.setEffect(null);
			});

			nodeBlur.setEffect(blur);
			dialog.show();
		} catch (IOException e) {
			
			LOGGER.error(e.getMessage(), e);
		}
	}

	private void configuraListenert() {
		txtId.focusedProperty().addListener((o, oldVal, newVal) -> {
			if (oldVal) {
				if (validaCampos()) {
					try {
						consulta = null;
						txtNome.setText("");
						imagem.setImage(null);
						
						txtId.setUnFocusColor(Color.web("#106ebe"));
						
						Long numero = Long.valueOf(getId().replaceAll("/[^0-9]+/g", ""));
						mal.getMyanimelist().clear();
						consulta = mal.addRegistro("", numero, false);
						ProcessaComicInfo.getById(numero, consulta);
						txtNome.setText(consulta.getNome());
						imagem.setImage(consulta.getImagem().getImage());
					} catch (Exception e) {
						
						LOGGER.error(e.getMessage(), e);
					}
				}
			}
		});
		
		txtId.setOnKeyPressed(new EventHandler<KeyEvent>() {
			@Override
			public void handle(KeyEvent ke) {
				if (ke.getCode().equals(KeyCode.ENTER))
					robot.keyPress(KeyCode.TAB);
			}
		});

		txtNome.setOnKeyPressed(new EventHandler<KeyEvent>() {
			@Override
			public void handle(KeyEvent ke) {
				if (ke.getCode().equals(KeyCode.ENTER))
					robot.keyPress(KeyCode.TAB);
			}
		});
	}

	@Override
	public void initialize(URL location, ResourceBundle resources) {
		configuraListenert();
	}

	public static URL getFxmlLocate() {
		return MangasComicInfoMalId.class.getResource("/view/Mangas/MangaComicInfoMalId.fxml");
	}
}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/controller/mangas/MangasComicInfoMalId.kt
package br.com.fenix.processatexto.controller.mangas

import br.com.fenix.processatexto.components.notification.AlertasPopup
import br.com.fenix.processatexto.model.entities.comicinfo.MAL
import br.com.fenix.processatexto.processar.comicinfo.ProcessaComicInfo
import com.jfoenix.controls.JFXButton
import com.jfoenix.controls.JFXDialog
import com.jfoenix.controls.JFXDialogLayout
import com.jfoenix.controls.JFXTextField
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.fxml.Initializable
import javafx.geometry.Insets
import javafx.scene.Node
import javafx.scene.Parent
import javafx.scene.control.Label
import javafx.scene.effect.BoxBlur
import javafx.scene.image.ImageView
import javafx.scene.input.KeyCode
import javafx.scene.layout.StackPane
import javafx.scene.paint.Color
import javafx.scene.robot.Robot
import javafx.scene.text.Font
import javafx.util.Callback
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.URL
import java.util.*


class MangasComicInfoMalId : Initializable {

    @FXML
    lateinit var txtId: JFXTextField

    @FXML
    lateinit var txtNome: JFXTextField

    @FXML
    lateinit var imagem: ImageView

    private val mal: MAL = MAL("", "")
    private val robot: Robot = Robot()
    private lateinit var consulta: MAL.Registro

    private fun validaCampos(): Boolean {
        val valida: Boolean = txtId.text.replace("/\\D+/g".toRegex(), "").trim().isNotEmpty()
        if (!valida)
            txtId.unFocusColor = Color.RED
        return valida
    }

    val id: String
        get() = if (validaCampos()) txtId.text else ""
    val objeto: MAL.Registro
        get() = consulta

    private fun configuraListenert() {
        txtId.focusedProperty().addListener { _, oldVal, _ ->
            if (oldVal) {
                if (validaCampos()) {
                    try {
                        txtNome.text = ""
                        imagem.image = null
                        txtId.unFocusColor = Color.web("#106ebe")
                        val numero = id.replace("/\\D+/g".toRegex(), "").toLong()
                        mal.myanimelist.clear()
                        consulta = mal.addRegistro("", numero, false)
                        ProcessaComicInfo.getById(numero, consulta)
                        txtNome.text = consulta.nome
                        imagem.image = consulta.imagem!!.image
                    } catch (e: Exception) {
                        LOGGER.error(e.message, e)
                    }
                }
            }
        }
        txtId.setOnKeyPressed { ke -> if (ke.code.equals(KeyCode.ENTER)) robot.keyPress(KeyCode.TAB) }
        txtNome.setOnKeyPressed { ke -> if (ke.code.equals(KeyCode.ENTER)) robot.keyPress(KeyCode.TAB) }
    }

    @Override
    override fun initialize(location: URL?, resources: ResourceBundle?) {
        configuraListenert()
    }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(MangasComicInfoMalId::class.java)

        private val STYLE_SHEET: String = AlertasPopup::class.java.getResource("/css/Dark_Theme.css").toExternalForm()
        private lateinit var btnConfirmar: JFXButton
        private lateinit var btnVoltar: JFXButton
        private lateinit var dialog: JFXDialog
        fun abreTelaCorrecao(rootStackPane: StackPane, nodeBlur: Node, callback: Callback<MAL.Registro, Boolean>) {
            try {
                val blur = BoxBlur(3.0, 3.0, 3)
                val dialogLayout = JFXDialogLayout()
                dialog = JFXDialog(rootStackPane, dialogLayout, JFXDialog.DialogTransition.CENTER)
                val loader = FXMLLoader()
                loader.location = fxmlLocate
                val newAnchorPane: Parent = loader.load()
                val cnt: MangasComicInfoMalId = loader.getController()
                val titulo = Label("Troca de Id")
                titulo.font = Font.font(20.0)
                titulo.textFill = Color.web("#ffffff", 0.8)
                val botoes = mutableListOf<JFXButton>()
                btnConfirmar = JFXButton("Confirmar")
                btnConfirmar.setOnAction {
                    val numero: String = cnt.id.replace("/\\D+/g".toRegex(), "")
                    if (numero.isNotEmpty()) {
                        callback.call(cnt.objeto)
                        dialog.close()
                    }
                }
                btnConfirmar.styleClass.add("background-Green2")
                botoes.add(btnConfirmar)
                btnVoltar = JFXButton("Voltar")
                btnVoltar.setOnAction { dialog.close() }
                btnVoltar.styleClass.add("background-White1")
                botoes.add(btnVoltar)
                dialogLayout.setHeading(titulo)
                dialogLayout.setBody(newAnchorPane)
                dialogLayout.setActions(botoes)
                dialog.stylesheets.add(STYLE_SHEET)
                dialog.padding = Insets(0.0, 0.0, 0.0, 0.0)
                dialog.setOnDialogClosed { nodeBlur.effect = null }
                nodeBlur.effect = blur
                dialog.show()
            } catch (e: IOException) {
                LOGGER.error(e.message, e)
            }
        }

        val fxmlLocate: URL get() = MangasComicInfoMalId::class.java.getResource("/view/mangas/MangaComicInfoMalId.fxml") as URL
    }
}